akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -699,18 +790,63 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
    *
    * This object can be cached and reused within the same [[GraphStageLogic]].
    */
-  final def getAsyncCallback[T](handler: T ⇒ Unit): AsyncCallback[T] = {
+  final protected def getAsyncCallback[T](handler: T ⇒ Unit): AsyncCallback[T] = {
     new AsyncCallback[T] {
       override def invoke(event: T): Unit =
         interpreter.onAsyncInput(GraphStageLogic.this, event, handler.asInstanceOf[Any ⇒ Unit])
     }
   }
 
+  // TODO rethink if it's worth it; move the Updater to an object; etc
+  // via atomics since users may do something weird - try to Future { init } or something...
+  private final val _setStageActorRef = AtomicReferenceFieldUpdater.newUpdater(classOf[GraphStageLogic], classOf[StageActorRef], "_stageActorRefDoNotCallMeDirectly")
+  @volatile private var _stageActorRefDoNotCallMeDirectly: StageActorRef = _ // DO NOT TOUCH!
+  final implicit def stageActorRef: ActorRef =
+    if (_stageActorRefDoNotCallMeDirectly ne null) _stageActorRefDoNotCallMeDirectly // TODO rethink this
+    else throw StageActorRefNotInitializedException()
+
+  case class StageActorRefNotInitializedException() extends RuntimeException("You must first call getStageActorRef, to initialize the Actors behaviour")
+
+  /**
+   * Initialize a [[StageActorRef]] which can be used to interact with from the outside world "as-if" an [[Actor]].
+   * The messages are looped through the [[getAsyncCallback]] mechanism of [[GraphStage]] so they are safe to modify
+   * internal state of this stage.
+   *
+   * This method must not (the earliest) be called after the [[GraphStageLogic]] constructor has finished running,
+   * for example from the [[preStart]] callback the graph stage logic provides.
+   *
+   * Created [[StageActorRef]] to get messages and watch other actors in synchronous way.
+   *
+   * The [[StageActorRef]]'s lifecycle is bound to the Stage, in other words when the Stage is finished,
+   * the Actor will be terminated as well. The entity backing the [[StageActorRef]] is not a real Actor,
+   * but the [[GraphStageLogic]] itself, therefore it does not react to [[PoisonPill]].
+   *
+   * @param receive callback that will be called upon receiving of a message by this special Actor
+   * @return minimal actor with watch method
+   */
+  // FIXME: I don't like the Pair allocation :(
+  final protected def initStageActorRef(receive: ((ActorRef, Any)) ⇒ Unit): StageActorRef = {
```

Comment:
ok renamed right away then (I liked init more somehow, but not married to the name ;))